### ISSUE TITLE
Resolution of single-label names via quad-100 is laggy for Windows

### ISSUE THREAD

**Issue by danderson** [tailscale/tailscale#1659]
Found while implementing split DNS on Windows.

Setup: enable MagicDNS on your domain, configure a resolver other than your LAN's resolver (e.g. 8.8.8.8), add search domains other than MagicDNS (e.g. tailscale.com), enable split DNS mode. Ensure your LAN resolver also sets some search domain, other than the ones set by Tailscale (e.g. universe.tf).

Dataflow: requests for `<domain>.beta.tailscale.net` and `tailscale.com` flow to the 100.100.100.100 resolver on localhost. `<domain>.beta.tailscale.net` gets handled by the MagicDNS hosts table, `tailscale.com` gets forwarded to 8.8.8.8. Everything else falls through to the LAN resolvers.

Open a command prompt, and start pinging single-label things whose names resolve through these different paths:
 - `ping go` -> `go.universe.tf`, resolution is ~instant via LAN resolvers and pings begin.
 - `ping derp1` -> `derp1.tailscale.com`, resolution via 100.100.100.100 forwarded to 8.8.8.8, but resolution stalls for 1-2s before pings begin.
 - `ping acrux` (a MagicDNS name) -> `acrux.natulte.net.beta.tailscale.net`, but resolution stalls for 1-2s before pings begin.

Additional spam logging in tailscaled shows MagicDNS receiving A and AAAA queries promptly, and returning responses promptly (either forwarded from 8.8.8.8, or from its internal hosts table). So, seemingly the delay is not in tailscaled itself, but some external process.

Separately, Wireshark shows Windows sending LLMNR queries to the LAN, simultaneously with the MagicDNS queries. It does _not_ send LLMNR requests for names that resolve through the regular LAN resolver, only for our queries.

Hypothesis: waiting on LLMNR timeout is what's slowing us down.

Question: why do only our queries get LLMNR'd? What's different in our config vs. the default DNS config?

Rant: Microsoft does not provide documentation on the algorithm used for resolving single-label DNS names, so we're left guessing.

Datapoint: I turned off LLMNR on the Tailscale network adapter (since it can't multicast anyway). This didn't change anything.

**Comment by danderson** [comment]
@bradfitz remembers that Perkeep once hit something about maybe SOA records relating to custom DNS resolvers, and will add details here once dug up.

**Comment by danderson** [comment]
Also, disabling LLMNR globally didn't help, so it seems to be something else.

**Comment by bradfitz** [comment]
@danderson, the DNS thing from Perkeep I was remembering was https://github.com/perkeep/perkeep/commit/0273e6c5f54c44bc76055748b70076c0d9ce68a6

**Comment by mayakacz** [comment]
Would like to confirm if this is still happening.

**Comment by bradfitz** [comment]
@dblohm7, do you know anything about such Windows LLMNR things?

**Comment by dblohm7** [comment]
Not off the top of my head, but I'll do some spelunking and see what I can come up with.

**Comment by DentonGentry** [comment]
On a Windows 10 system running Tailscale 1.28 when I ping a single-label MagicDNS name, in Wireshark on the Tailscale interface I see:
- regular DNS requests, which receive responses ~instantly
- mDNS requests which get no response, followed by a retransmitted request one second later
- pings start 2.26 seconds after the initial DNS query which was answered immediately

[mdns.windows.pcapng.zip](https://github.com/tailscale/tailscale/files/9248143/mdns.windows.pcapng.zip)

**Comment by DentonGentry** [comment]
There are registry keys to disable mDNS globally: https://f20.be/blog/mdns
So far I haven't found a setting to disable it for a specific interface.

PPP interfaces have a `connection.mdns` setting, but it appears to be some kind of dialing script: https://docs.microsoft.com/en-us/answers/questions/207832/how-i-can-setup-some-additional-parameters-of-pppo.html

**Comment by DentonGentry** [comment]
Disappointingly, we already set a registry key to disable multicast on the Tailscale interface but the mDNS frames are sent regardless.
<img width="769" alt="Screen Shot 2022-08-02 at 11 13 42 PM" src="https://user-images.githubusercontent.com/1140756/182537127-89203cdd-024d-4f35-aa78-78ba1fa877fa.png">

**Comment by DentonGentry** [comment]
Other registry settings tried with no effect:
- `IGMPLevel` dword32 = 0
- `EnableWsd` dword32 = 0
- `EnableMDNS` dword32 = 0 (which exists in HKLM\CurrentControlSet\Services\Dnscache\Parameters but I hoped there was a per-interface setting)

**Comment by DentonGentry** [comment]
[This article](https://f20.be/blog/mdns) is pretty convincing that there is not an interface-specific registry key to not send mDNS requests. They used Procmon from [Sysinternals](https://docs.microsoft.com/nb-no/sysinternals/downloads/) to trace the DnsCache process. It checks:
- `HKLM\SOFTWARE\Policies\Microsoft\Windows NT\DNSClient\EnableMDNS`
- `HKLM\System\CurrentControlSet\Services\Dnscache\Parameters\EnableMDNS`

HOWEVER: I disabled mDNS system-wide using `HKLM\System\CurrentControlSet\Services\Dnscache\Parameters\EnableMDNS`, and it did work. There is no mDNS packet on the Tailscale interface now. Yet the first ping is still delayed by 2.3 seconds.

[b1659 disabled mDNS.pcapng.zip](https://github.com/tailscale/tailscale/files/9251780/b1659.disabled.mDNS.pcapng.zip)

The delay is apparently not mDNS, it is something else.

**Comment by dblohm7** [comment]
@DentonGentry Here's a shot in the dark, but what the hell: I took a look a that pcap. I noticed that shortly after the delay there's some activity surrounding detection of proxy auto-config data.

I also noticed this warning in the docs for [`WinHttpDetectAutoProxyConfigUrl`](https://docs.microsoft.com/en-us/windows/win32/api/winhttp/nf-winhttp-winhttpdetectautoproxyconfigurl):

> Note that because the `WinHttpDetectAutoProxyConfigUrl` function takes time to complete its operation, it should not be called from a UI thread.

If it is slow enough to be noticeable on a UI thread, then it's probably slow enough to be responsible for a delay such as this. I'm wondering if it's doing something stupid like timing out trying to resolve the URL via DHCP even though DHCP is not present on our interface. Worth experimenting, perhaps?

**Comment by DentonGentry** [comment]
That is a good thought. However after adding `HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Internet Settings\WinHttp\DisableWpad` with a value of 0x1 I see no further lookups for `wpad` but the first ping is still delayed by 2.3 seconds.

[b1659 disabled wpad.pcapng.zip](https://github.com/tailscale/tailscale/files/9265238/b1659.disabled.wpad.pcapng.zip)

**Comment by DentonGentry** [comment]
There is exactly one search result for: _windows "first ping" delayed "2.3 seconds"_

https://www.google.com/search?q=windows+%22first+ping%22+delayed+%222.3+seconds%22

It is from cern.ch, and looks very promising: _"Then the second and third ping are extracted at 12.295 and 22.545 seconds after T5 respectively. In between them, at 19.29 seconds from T5 there is a 2.3 seconds long slow spill."_

The article was written at CERN in 1989 — which substantially predates Windows 10, but CERN is frequently ahead of its time. The title of the article is "NEUTRINO PRODUCTION OF OPPOSITE SIGN DIMUONS AT THE FNAL TEVATRON"

Apparently our problem is neutrinos. We either need more neutrinos, or less neutrinos. I'm not quite sure.

**Comment by maisem** [comment]
Another theory is that Windows is somehow prioritizing search domains, and for unqualified DNS names like `derp1` or `acrux` it is waiting to see if it ever gets a response from a higher priority search domain.

**Comment by maisem** [comment]
This looks promising, I made the `foo.ts.net` search domain the only one and it immediately starts the ping

**Comment by maisem** [comment]
I spent the last two days digging into this. This looks more and more like a bug in Windows.

[Currently](https://github.com/tailscale/tailscale/blob/539c5e44c56e9240bb5a88422eee4b7fc4966712/net/dns/manager_windows.go#L232-L241) in Split DNS mode, Tailscale does not set the DNS server on the Tailscale interface. It instead uses NRPT to configure the routes. For the context of this bug, this means that the Tailscale DNS entries get lower priority than the ones configured on the LAN interface.

I manually set the DNS entry on the Tailscale interface to see how it behaves if Tailscale has the priority
<details>

```
PS C:\WINDOWS\system32> netsh int ip show dns

Configuration for interface "Tailscale"
    Statically Configured DNS Servers:    100.100.100.100
    Register with which suffix:           Primary only

Configuration for interface "Ethernet 2"
    DNS servers configured through DHCP:  10.0.6.1
    Register with which suffix:           None

Configuration for interface "Loopback Pseudo-Interface 1"
    Statically Configured DNS Servers:    None
    Register with which suffix:           Primary only

PS C:\WINDOWS\system32> netsh int ip set dns Tailscale static 100.100.100.100

PS C:\WINDOWS\system32> netsh int ip show dns

Configuration for interface "Tailscale"
    Statically Configured DNS Servers:    100.100.100.100
    Register with which suffix:           Primary only

Configuration for interface "Ethernet 2"
    DNS servers configured through DHCP:  10.0.6.1
    Register with which suffix:           None

Configuration for interface "Loopback Pseudo-Interface 1"
    Statically Configured DNS Servers:    None
    Register with which suffix:           Primary only

```
</details>


A very minimal repro 

```go
package main

import (
	"context"
	"fmt"
	"net"
	"time"
)

func main() {
	r := net.Resolver{}
	ctx := context.Background()
	// "mon" is resolvable over Tailscale, "hl2280" is resolvable over local DNS
	for _, h := range []string{"mon", "hl2280"} {
		t := time.Now()
		addrs, err := r.LookupHost(ctx, h)
		dur := time.Since(t)
		fmt.Printf("looking up %q took %v: %v, %v\n", h, dur, addrs, err)
	}
}
```
With Tailscale as the primary
```
PS C:\Users\maise\Desktop\dev\dns> go run .
looking up "mon" took 3.471ms: [100.X.Y.Z], <nil>
looking up "hl2280" took 2.3446517s: [10.X.Y.Z], <nil>
```
With LAN as the primary
```
PS C:\Users\maise\Desktop\dev\dns> go run .
looking up "mon" took 2.278993s: [100.X.Y.Z], <nil>
looking up "hl2280" took 1.4715ms: [10.X.Y.Z], <nil>
```

Note that just adding the DNS to the Tailscale interface is sufficient to give it priority as the Tailscale interface metric is lower than the ethernet interface metric, making the Tailscale interface metric to be higher makes the other one primary.
<details>

```

PS C:\WINDOWS\system32> Get-NetIPInterface

ifIndex InterfaceAlias                  AddressFamily NlMtu(Bytes) InterfaceMetric Dhcp     ConnectionState PolicyStore
------- --------------                  ------------- ------------ --------------- ----     --------------- -----------
19      Ethernet 2                      IPv6                  1500              25 Enabled  Connected       ActiveStore
15      Tailscale                       IPv6                  1420               5 Disabled Connected       ActiveStore
1       Loopback Pseudo-Interface 1     IPv6            4294967295              75 Disabled Connected       ActiveStore
19      Ethernet 2                      IPv4                  1500              25 Enabled  Connected       ActiveStore
15      Tailscale                       IPv4                  1280               5 Disabled Connected       ActiveStore
1       Loopback Pseudo-Interface 1     IPv4            4294967295              75 Disabled Connected       ActiveStore

PS C:\WINDOWS\system32> Set-NetIPInterface -InterfaceIndex 15 -InterfaceMetric 26

PS C:\WINDOWS\system32> Get-NetIPInterface

ifIndex InterfaceAlias                  AddressFamily NlMtu(Bytes) InterfaceMetric Dhcp     ConnectionState PolicyStore
------- --------------                  ------------- ------------ --------------- ----     --------------- -----------
19      Ethernet 2                      IPv6                  1500              25 Enabled  Connected       ActiveStore
15      Tailscale                       IPv6                  1420              26 Disabled Connected       ActiveStore
1       Loopback Pseudo-Interface 1     IPv6            4294967295              75 Disabled Connected       ActiveStore
19      Ethernet 2                      IPv4                  1500              25 Enabled  Connected       ActiveStore
15      Tailscale                       IPv4                  1280              26 Disabled Connected       ActiveStore
1       Loopback Pseudo-Interface 1     IPv4            4294967295              75 Disabled Connected       ActiveStore


```
</details>

**Comment by maisem** [comment]
Turns out setting `HKLM\SOFTWARE\Policies\Microsoft\Windows NT\DNSClient\EnableMDNS` has immediate impact as opposed to `HKLM\SYSTEM\CurrentControlSet\Services\Dnscache\Parameters\EnableMDNS` which requires a reboot

**Comment by louislam** [comment]
I think it is not solved yet, I am still experiencing slow resolution of the first query. 

Windows 21H2 (OS Build 22000.856)
Tailscale Version 1.30.2 (t24c524c78-gc399ae6fa)
